pvlib/data/gold/sdm.py

```diff
@@ -0,0 +1,832 @@
+    from interval import interval, imath
+
+    # Approximate largest possible absolute residual value, determined
+    #  empirically
+    tol = 1.e-12
+
+    tol_interval = interval([-tol, tol])
+
+    # Generate I-V curves, each with same number of points, over matrix of
+    #  (G,T) combinations
+    G_list = [100., 200., 400., 600., 800., 1000., 1100.]  # W/m^2
+    T_list = [15., 25., 50., 75.]  # degC
+    NUM_PTS = 21
+
+    # The "base" devices, having ideal, normal, and degraded versions
+    devices = [
+        bootstrap_si_device_bishop(),
+        bootstrap_si_device_bishop(N_s=8),
+        bootstrap_si_device_pvmismatch(),
+        pvsystem.retrieve_sam('cecmod').SunPower_SPR_E20_327,
+        pvsystem.retrieve_sam('cecmod').First_Solar_FS_495
+    ]
+
+    # Initialize the output, recording metadata
+    gold_dataset = {
+        "timestamp_utc_iso": datetime.datetime.utcnow().isoformat(),
+        "dataset_version": '1.0.0',  # dataset follows SemVer
+        "python_version": sys.version,
+        "numpy_version": numpy.__version__,
+        "pvlib_version": pvlib.version.__version__,
+        "platform_version": platform.platform(),
+        "residual_interval_tolerance": tol,
+        "units": {
+            "G": 'W/m^2',
+            "T": 'degC',
+            "i_l": 'A',
+            "i_0": 'A',
+            "r_s": 'Ohm',
+            "g_sh": 'S',
+            "nNsVth": 'V',
+            "v_gold": 'V',
+            "i_gold": 'A'
+        },
+        "devices": []
+    }
+
+    # For each ideal, nominal, or degraded device, create matrix of I-V curves
+    for device in devices:
+        device_dict = {
+            "name": device.name,
+            "Technology": device.Technology,
+            "N_s": device.N_s,
+            "iv_curves": []
+        }
+
+        # Choose bandgap parameters based on device technology
+        # Add new technologies if/when needed, should match a CEC module
+        #  database type
+        # These are NOT saved, but are versioned with the dataset version
+        if device.Technology == 'Mono-c-Si' or \
+           device.Technology == 'Multi-c-Si':
+            EgRef = 1.121
+            dEgdT = -0.0002677
+        elif device.Technology == 'CdTe':
+            EgRef = 1.475
+            dEgdT = -0.0003
+        else:
+            raise ValueError('Unsupported device technology.')
+
+        # Do all "flavor" combinations of ideal device, nominal device, and
+        #  degraded device
+        r_s_scale_grid, g_sh_scale_grid = numpy.meshgrid([0., 1., 5.],
+                                                         [0., 1., 5.])
+
+        # For all "flavors" of this device, create I-V curves over (G,T) matrix
+        for r_s_scale, g_sh_scale in zip(r_s_scale_grid.flat,
+                                         g_sh_scale_grid.flat):
+            for G in G_list:
+                for T in T_list:
+                    # Compute the model parameters at each
```

Comment:
This line through 373 should be its own function e.g. ``create_device_dict(device, G, T, r_s_scale, g_sh_scale)``. Either a public or private function is fine with me here.